lens/private/list/flatten.rkt

```diff
@@ -0,0 +1,204 @@
+(define (flatten/depth n lst*)
+  (cond [(zero? n) (list lst*)]
+        [else (append*n (sub1 n) lst*)]))
+
+;; unflatten/depth : n (Listof* A n) (Listof B) -> (Listof* B n)
+(define (unflatten/depth n lst* lst)
+  (cond [(zero? n)
+         (match-define (list v) lst)
+         v]
+        [else
+         (unappend*n (sub1 n) lst* lst)]))
+
+;; append*n : n (Listof (Listof* A n)) -> (Listof A)
+(define (append*n n lst*)
+  (cond [(zero? n) lst*]
+        [else (append*n (sub1 n) (append* lst*))]))
+
+;; unappend*n : n (Listof (Listof* A n)) (Listof B) -> (Listof (Listof* B n))
+(define (unappend*n n lst* lst)
+  (cond [(zero? n) lst]
+        [else (restore-structure
+               lst*
+               (unappend*n (sub1 n) (append* lst*) lst))]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(module+ test
+  (test-case "append*-lens"
+    (check-equal? (lens-view append*-lens (list (list 1) (list 2 3) (list)))
+                  (list 1 2 3))
+    (check-equal? (lens-set append*-lens
+                            (list (list 1) (list 2 3) (list))
+                            (list 'a 'b 'c))
+                  (list (list 'a) (list 'b 'c) (list)))
+
+    (check-equal? (lens-transform append*-lens
+                                  (list (list 1) (list 2 3) (list))
+                                  (lambda (lst)
+                                    ;; a length-preserving computation
+                                    (let loop ([acc (list)] [sum 0] [lst lst])
+                                      (match lst
+                                        [(list) (reverse acc)]
+                                        [(cons fst rst)
+                                         (loop (cons (+ sum fst) acc)
+                                               (+ sum fst)
+                                               rst)]))))
+                  (list (list 1) (list 3 6) (list)))
+
+    (check-equal? (lens-transform append*-lens
+                                  (list (list #'(+ a)) (list #'(- a b) #'(* c d)) (list))
+                                  (lambda (lst)
+                                    ;; a length-preserving computation
+                                    (syntax-parse
+                                        (expand #`(#%expression (λ (a b c d) (#%app list #,@lst))))
+                                      #:literals (#%plain-lambda #%plain-app list)
+                                      [(#%expression (#%plain-lambda (x ...) (#%plain-app list e ...)))
+                                       (syntax->datum #'[e ...])])))
+                  (list (list '(#%app + a))
+                        (list '(#%app - a b) '(#%app * c d))
+                        (list)))
+
+    (test-lens-laws append*-lens
+                    (list (list 1) (list 2 3) (list))
+                    (list 'a 'b 'c)
+                    (list "a" "b" "c"))
+    )
+  
+  (test-case "flatten/depth-lens"
```

Comment:
Could this be split into smaller test cases with more informative test descriptions?